src/unix/exec.c

```diff
@@ -326,24 +409,131 @@ process exec_elf(buffer ex, process kp)
+    exec_debug("%s: curr %ld, readlen %ld\n", __func__, curr, readlen);
+    filesystem_read_linear(f, buffer_ref(b, curr), irangel(curr, readlen), self);
+    return true;
 }
 
+closure_function(4, 2, void, exec_elf_read,
+                 fsfile, f, buffer, b, process, kp, status_handler, complete,
+                 status, s, bytes, length)
+{
+    fsfile f = bound(f);
+    buffer b = bound(b);
+    process kp = bound(kp);
+    status_handler complete = bound(complete);
+    exec_debug("%s: status %v, read %ld bytes\n", __func__, s, length);
+    if (!is_ok(s)) {
+        apply(complete, timm_up(s, "result", "failed to read elf file"));
+        closure_finish();
+        return;
+    }
+    buffer_produce(b, length);
+    exec_debug("buffer length %ld\n", buffer_length(b));
+    Elf64_Ehdr *e = (Elf64_Ehdr *)buffer_ref(b, 0);
+    const char *interp_path = 0;
+    boolean ingest_symbols = !!get(kp->process_root, sym(ingest_program_symbols));
+
+    /* This procedure will repeat until we have all the ELF info we need. */
+    range load_range = irange(infinity, 0);
+    foreach_phdr(e, p) {
+        if (p->p_type == PT_INTERP) {
+            interp_path = (void *)e + p->p_offset;
+            exec_debug("interp offset p->p_offset %ld, p->p_filesz %ld\n", p->p_offset, p->p_filesz);
+            if (elf_check_extend(p->p_offset + p->p_filesz, f, b, (io_status_handler)closure_self()))
+                return;
+        } else if (p->p_type == PT_LOAD) {
+            if (p->p_vaddr < load_range.start)
+                load_range.start = p->p_vaddr;
+            u64 segend = p->p_vaddr + p->p_memsz;
+            if (segend > load_range.end)
+                load_range.end = segend;
+        }
+    }
+    if (ingest_symbols) {
+        if (elf_check_extend(e->e_shoff + (e->e_shnum * e->e_shentsize), f, b,
+                             (io_status_handler)closure_self()))
+            return;
+        u64 req_size = 0;
+        foreach_shdr(e, shdr) {
+            if (shdr->sh_type == SHT_SYMTAB || shdr->sh_type == SHT_STRTAB) {
+                u64 sz = shdr->sh_offset + shdr->sh_size;
+                if (sz > req_size)
+                    req_size = sz;
+            }
+        }
+        if (elf_check_extend(req_size, f, b, (io_status_handler)closure_self()))
+            return;
+    }
+    exec_debug("finished reading ELF data, buffer length %ld\n", buffer_length(b));
+    closure_finish();
+    exec_elf_finish(b, f, kp, load_range, interp_path, ingest_symbols, complete);
+}
+
+void exec_elf(process kp, string program_path, status_handler complete)
+{
+    exec_debug("%s: path \"%b\", complete %p (%F)\n", __func__, program_path, complete, complete);
+    kernel_heaps kh = (kernel_heaps)(kp->uh);
+    heap general = heap_locked(kh);
+    tuple root = kp->process_root;
+    fsfile f = fsfile_open_or_create(program_path, false);
+    if (!f) {
+        apply(complete, timm("result", "unable to open program file %v", program_path));
```

Comment:
Using fsfile_open_or_create() means that if the file doesn't exist it will be created (with length 0), so instead of an "unable to open program file" error message there is a more cryptic "failed to parse elf file" message. The same goes for the interpreter, if the file doesn't exist we get a "failed to parse elf file" message instead of "couldn't find program interpreter".
Perhaps we could change fsfile_open_or_create() to a more generic fsfile_open() that takes an additional boolean argument indicating whether the file should be created if not found.

Reply:
Ok, I added an fsfile_open() function for opening existing files, but left fsfile_open_or_create() in place for existing uses. I also noticed that, prior to this PR, a missing program file would result in a similarly-cryptic message "cannot set on unknown value" (or an "assert(0)" failure and backtrace prior to new-tag-types merge) due to a set on a null tuple in startup(). I changed this to print an "unable to resolve program path" error if resolve_path() returns null.